main.cc

```diff
@@ -1282,7 +1282,7 @@ int main(int ac, char** av) {
                 api::unset_transport_controller(ctx).get();
             });
 
-            ::thrift_controller thrift_ctl(db, auth_service, qp);
+            ::thrift_controller thrift_ctl(db, auth_service, qp, service_memory_limiter);
```

Comment:
Commenting on this line just because I can't comment on the commit message :-(
In the commit message you say that "This mechanism heavily relies on the fact that there are zero preemption points between overwriting the service permit and reading it back by the handler.". However, I couldn't figure out where in the actual code this assumption is made. Could we have a FIXME in the actual code that assumes this?

Reply:
This assumption is a little convoluted, because it's spread over 4 separate entities - the Thrift server, the handler factory, the handler, and then the code which handles a single request; but I'll try to put a comment somewhere - it won't really be a FIXME, because I have no idea how it should be "fixed". The idea is that once a connection reads a request from its socket, it creates a service permit object which is used for reserving resources. Then, this connection puts this service permit in a special place in the **server** object - this one is tricky, because multiple connections share a single server - and then the code which handles a request gets the permit back from this special place in the server, which is global for all connections. And the assumption is that there is no preemption point between putting the service permit in this special place and retrieving it in the code which runs the request. Otherwise we'd have races, because this address is effectively shared between connections. And all that is just because the Thrift API doesn't offer any easy way of passing custom context to a single request handler, or at least I wasn't able to find such an API. A potentially safer alternative to that would be if the server kept a map of <connection, service_permit>, which would keep a separate place to pass the permit per each connection separately. That's additional overhead though.

Reply:
So even if not a map, maybe at least check if the server already has a permit set - and if it is, log an internal error? This will allow us to find such a bug if it ever happens because of future code changes. It will also give you a good place to put a FIXME (in front of this logging code) and the FIXME can include this explanation which you wrote here because obviously in a year nobody will remember why you made this assumption.

Reply:
Sure, but I'd limit this error to debug/trace, because I'd actually expect it to fire for cql-over-thrift requests (of which I wasn't even aware until I touched this part of our code base), because this series doesn't address this part of Thrift yet.